csrc/cutlass_extensions/vllm_numeric_conversion.cuh

```diff
@@ -140,6 +149,142 @@ struct ArrayConverterPacked32Bit {
+      static constexpr uint32_t NEG_INT8_REG2 = 0xFFFEFDFC;  // [-4,-3,-2,-1]
+
+      const int iters = std::max(PackedResultType::kElements / 4, size_t(1));
+      CUTLASS_PRAGMA_UNROLL
+      for (int ii = 0; ii < iters; ++ii, lut_idx >>= 16, sign >>= 16) {
+        uint32_t final_prmt_idx = final_prmt_base | sign;
+
+        // This uses a look up table to convert packed int4s to packed fp8s,
+        // using the int4 value as the index to prmt. It first select both the
+        // positive and negative candidates, then uses the sign bit to select
+        // the correct candidate.
+        asm volatile(
+            "{\n"
+            "  .reg .b32 pos_f8s, neg_f8s;\n"
+            "  prmt.b32 pos_f8s, %1, %2, %5;\n"
+            "  prmt.b32 neg_f8s, %3, %4, %5;\n"
+            "  prmt.b32 %0, neg_f8s, pos_f8s, %6;\n"
+            "}\n"
+            : "=r"(r[ii])
+            : "n"(POS_INT8_REG1), "n"(POS_INT8_REG2), "n"(NEG_INT8_REG1),
+              "n"(NEG_INT8_REG2), "r"(lut_idx), "r"(final_prmt_idx));
+      }
+
+      return reinterpret_cast<PackedResultType&>(r);
+    };
+  };
+
+ public:
+  CUTLASS_DEVICE
+  static result_type convert(source_type const& source) {
+    return ArrayConverterPacked32Bit<RegConvert, typename result_type::Element,
+                                     typename source_type::Element,
+                                     N>::convert(source);
+  }
+
+  CUTLASS_DEVICE
+  result_type operator()(source_type const& s) const { return convert(s); }
+};
+
+// for Array<cutlass::float_e4m3_t, N> <= Array<vllm_uint4b8_t, N>
+template <FloatRoundStyle Round, int N>
+struct NumericArrayConverter<cutlass::float_e4m3_t, vllm_uint4b8_t, N, Round> {
+  using result_type = Array<cutlass::float_e4m3_t, N>;
+  using source_type = Array<vllm_uint4b8_t, N>;
+
+  static FloatRoundStyle const round_style = Round;
+
+ private:
+  struct RegConvert {
+    template <typename PackedResultType>
+    CUTLASS_DEVICE static PackedResultType convert(Array<uint32_t, 1> src_) {
+      cutlass::AlignedArray<uint32_t, std::max(PackedResultType::kElements / 4,
+                                               size_t(1))>
+          r;
+      uint32_t src = src_[0];
+      // Determines if to get from the signed or unsigned candidates
+      uint32_t sign = (src & 0x88888888) >> 1;
+
+      // Ignore sign bit when indexing into LUT
+      uint32_t lut_idx = (src & 0x77777777);
+
+      // Signed is OR'd with 0x31203120 to find the correct value in the LUT
+      // (selects correct positive or negative candidate)
+      const uint32_t final_prmt_base = 0x32103210;
+
+      static constexpr uint32_t POS_E4M3s_REG1 = 0x44403800;  // [0, 1, 2, 3]
+      static constexpr uint32_t POS_E4M3s_REG2 = 0x4E4C4A48;  // [4, 5, 6, 7]
+      static constexpr uint32_t NEG_E4M3s_REG1 = 0xCACCCED0;  // [-8,-7,-6,-5]
+      static constexpr uint32_t NEG_E4M3s_REG2 = 0xB8C0C4C8;  // [-4,-3,-2,-1]
```

Comment:
It looks like the int4 -> fp8 and int4 -> int8 converters the same except for these constants? If so, might be nice to factor these out. Not a big deal though, because it'd be kind of annoying to do.

Reply:
Good call, bit janky but refactored to `lut_4bit_to_8bit_convert`